crypto/params.c

```diff
@@ -211,6 +211,8 @@ int OSSL_PARAM_set_int32(OSSL_PARAM *p, int32_t val)
     p->return_size = 0;
     if (p->data_type == OSSL_PARAM_INTEGER) {
         p->return_size = sizeof(int32_t); /* Minimum expected size */
+        if (p->data == NULL)
```

Comment:
Err.. it depends on the data_size.. for the 64 bit 'case' down further you change the return_size..

This pattern applies to most other types dealing with multiple data_sizes.

Reply:
If this passes - then it indicates we are missing tests :)

Reply:
No, you misunderstand.  If `data` is NULL, then the responder can only tell what size it needs, independently of what the caller might assign, and the comment on the line above is pretty clear, the minimum size is `sizeof(int32_t)`.  `data == NULL` is the caller asking what size is expected, not trying to assert a size.

Reply:
What about having a reasonably sized buffer for `data` and mallocing to size if the returned value doesn't fit?  In most cases the output size should be limited.

Reply:
I would of thought it needed to be the max size if you are going to do this. I think of it as how much space you need to allocate to fit your data into.

Reply:
@slontis, The code for `set_int32` *allows* the caller to provide a larger buffer, and adapts accordingly.  This is relevant on big-endian systems partcularly.  So the *required* size for a `int32` is the size of a `int32`.

I insist.

Reply:
> What about having a reasonably sized buffer for `data` and mallocing to size if the returned value doesn't fit? In most cases the output size should be limited.

So are you basically saying that we should only support this mechanism for variable size parameters?  How do you plan on documenting that in `doc/man3/OSSL_PARAM.pod`?  After all, *all* parameters are potentially variable size.

Reply:
Leave things as there are here.  The params allow a buffer to be passed in and they will return the needed size if it is too small.  My suggestion is to pass a reasonably sized buffer in and avoid the malloc and double call most of the time.

Reply:
> The code for set_int32 allows the caller to provide a larger buffer, and adapts accordingly.

Not completely freely.  It is something I'd like to see in due course but it can wait.

Reply:
I'm mixing up the infrastructure and the calling of it.